### PR TITLE
SqlFragment.appendValue()

### PR DESCRIPTION
**Pull request by labkey-matthewb**
#### Rationale
New clearly named methods on SqlFragment should help avoid anti-patterns.  PR also integrates the new methods with SqlBuilder subclass.

#### Related Pull Requests
* 

#### Changes
*

### CODE REVIEW

File: api/src/org/labkey/api/data/SQLFragment.java [hunk 1]
```diff
@@ -368,27 +383,130 @@ public SQLFragment append(CharSequence csq, int start, int end)
...
+    }
 
-        append("'").append(id).append("'");
-        if (!StringUtils.containsAny(name,"*/'\"?"))
-            append("/* ").append(c.getName()).append(" */");
+    public SQLFragment appendValue(Boolean B, SqlDialect dialect)
+    {
+        if (null == B)
+            return appendNull();
+        getStringBuilder().append(B ? dialect.getBooleanTRUE() : dialect.getBooleanFALSE());
```
Comment (labkey-susanh) [inline api/src/org/labkey/api/data/SQLFragment.java:410]: Can `dialect` be null? If not, perhaps annotate as `@NotNull`?

File: api/src/org/labkey/api/data/SQLFragment.java [hunk 2]
```diff
@@ -368,27 +383,130 @@ public SQLFragment append(CharSequence csq, int start, int end)
...
+    public SQLFragment appendValue(Enum<?> e)
     {
-        getStringBuilder().append(o);
+        {
+            StringBuilder sb = getStringBuilder();
+            if (sb.length() > 0)
+            {
+                char last = sb.charAt(sb.length() - 1);
+                if (last == '\'')
+                    throw new IllegalStateException();
```
Comment (labkey-susanh) [inline api/src/org/labkey/api/data/SQLFragment.java]: Perhaps include a message about what kind of illegal state this is?

File: api/src/org/labkey/api/data/SQLFragment.java [hunk 3]
```diff
@@ -368,27 +383,130 @@ public SQLFragment append(CharSequence csq, int start, int end)
...
+            }
+        }
+        if (null == e)
+            return appendNull();
+        String name = e.name();
+        if (name.contains("'"))
+            throw new IllegalStateException();
```
Comment (labkey-susanh) [inline api/src/org/labkey/api/data/SQLFragment.java:488]: Ditto about a message here.

File: api/src/org/labkey/api/data/SQLFragment.java [hunk 4]
```diff
@@ -497,18 +614,28 @@ public SQLFragment append(char ch)
...
+        return appendValue(s, null);
+    }
+
+    public SQLFragment appendValue(CharSequence s, SqlDialect d)
+    {
+        if (null==s)
+            return appendNull();
+        if (null==d || s.length() > 200)
+            return append("?").add(s.toString());
+        getStringBuilder().append(d.getStringHandler().quoteStringLiteral(s.toString()));
```
Comment (labkey-susanh) [inline api/src/org/labkey/api/data/SQLFragment.java]: Could call `appendStringLiteral` here.

File: api/src/org/labkey/api/data/SqlExecutor.java [hunk 5]
```diff
@@ -157,7 +157,14 @@ public Integer execute(Connection conn, SqlDialect dialect, SQLFragment sqlFragm
...
                      Parameter.ParameterList jdbcParameters = new Parameter.ParameterList())
                 {
-                    Table.setParameters(stmt, parameters, jdbcParameters);
+                    try
+                    {
+                        Table.setParameters(stmt, parameters, jdbcParameters);
+                    }
+                    catch (Exception x)
+                    {
+                        x.printStackTrace();
```
Comment (labkey-susanh) [inline api/src/org/labkey/api/data/SqlExecutor.java]: We really want this to go to stderr? And continue on?

File: api/src/org/labkey/api/data/StatementUtils.java [hunk 6]
```diff
@@ -1189,6 +1188,9 @@ else if (jdbcType.isDecimal() && ph.getScale() > 0)
     }
 
 
+    /* We could use SQLFragment.appendValue() for most of these.  However, here it important to force
```
Comment (labkey-susanh) [inline api/src/org/labkey/api/data/StatementUtils.java]:
```suggestion
    /* We could use SQLFragment.appendValue() for most of these.  However, here it is important to force

```

File: api/src/org/labkey/api/query/FieldKey.java [hunk 7]
```diff
@@ -461,4 +461,10 @@ public void testRemap()
             }
         }
     }
+
+    @Override
+    public String toString()
```
Comment (labkey-susanh) [inline api/src/org/labkey/api/query/FieldKey.java]: Is this needed if it just calls `super.toString()`?

File: experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java [hunk 8]
```diff
@@ -3242,7 +3240,7 @@ private void syncInner(@Nullable Map<String, Integer> cpasTypeToObjectId)
                     .append("SELECT m.Container, m.LSID, m.CpasType, m.ObjectId, pa.CpasType AS pa_cpas_type FROM exp.material m\n")
                     .append("INNER JOIN exp.MaterialInput mi ON m.rowId = mi.materialId\n")
                     .append("INNER JOIN exp.ProtocolApplication pa ON mi.TargetApplicationId = pa.RowId\n")
-                    .append("WHERE pa.RunId = ").append(runId).append(" AND pa.CpasType IN ('").append(ExperimentRun.name()).append("','").append(ExperimentRunOutput.name()).append("')");
+                    .append("WHERE pa.RunId = ").appendValue(runId).append(" AND pa.CpasType IN (").appendValue(ExperimentRun).append(",").appendValue(ExperimentRunOutput.name()).append(")");
```
Comment (labkey-susanh) [inline experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java]:
```suggestion
                    .append("WHERE pa.RunId = ").appendValue(runId).append(" AND pa.CpasType IN (").appendValue(ExperimentRun).append(",").appendValue(ExperimentRunOutput).append(")");
```

File: query/src/org/labkey/query/sql/Method.java [hunk 9]
```diff
@@ -1402,19 +1407,18 @@ public SQLFragment getSQL(SqlDialect dialect, SQLFragment[] arguments)
 
 
 
-    public static SQLFragment escapeLikePattern(SQLFragment f, char escapeChar, @Nullable String prepend, @Nullable String append)
+    public static SQLFragment escapeLikePattern(SqlDialect d, SQLFragment f, char escapeChar, @Nullable String prepend, @Nullable String append)
```
Comment (labkey-susanh) [inline query/src/org/labkey/query/sql/Method.java]: Is `SqlDialect` nullable here?

File: query/src/org/labkey/query/sql/Method.java [hunk 10]
```diff
@@ -1423,32 +1427,39 @@ public static SQLFragment escapeLikePattern(SQLFragment f, char escapeChar, @Nul
...
         String s = f.getSQL();
-        if (s.length() < 2 || !s.startsWith("'"))
+        if ("?".equals(s) && f.getParams().size()==1)
+            return f.getParams().get(0) instanceof String;
+        if (f.getParams().size() > 0)
             return false;
-        return s.length()-1 == s.indexOf('\'',1);
+        if (s.length() >= 2 && s.startsWith("'"))
+            return s.length()-1 == s.indexOf('\'',1);
+        if (s.length() >= 3 && s.startsWith("N'"))
```
Comment (labkey-susanh) [inline query/src/org/labkey/query/sql/Method.java:1452]: What's this case for? Perhaps add a comment?

**Review by labkey-susanh — APPROVED** [review]

File: api/src/org/labkey/api/data/SQLFragment.java (hunk 2 above)
Comment (labkey-matthewb) [reply api/src/org/labkey/api/data/SQLFragment.java]: Should have removed that when I added the debug code to append(CharSequence)

File: api/src/org/labkey/api/data/SQLFragment.java (hunk 4 above)
Comment (labkey-matthewb) [reply api/src/org/labkey/api/data/SQLFragment.java]: I like that

File: api/src/org/labkey/api/data/SqlExecutor.java (hunk 5 above)
Comment (labkey-matthewb) [reply api/src/org/labkey/api/data/SqlExecutor.java]: Thank you, that is unrelated and got swept up by accident.

Comment (labkey-matthewb) [reply api/src/org/labkey/api/data/SqlExecutor.java]: I though I had reverted that.

**Review by labkey-klum — APPROVED** [review]

**Pull request merged**